File: solutions/haskell/proverb/2/src/Proverb.hs
module Proverb (recite) where

import Text.Printf (printf)

recite :: [String] -> String
recite [] = ""
recite pieces@(first : rest) =
  let
    verse = printf "For want of a %s the %s was lost."
    ending = printf "And all for the want of a %s."
   in
    init . unlines $ zipWith verse pieces rest ++ [ending first]

File: solutions/clojure/proverb/2/src/proverb.clj
(ns proverb)

(defn recite [items]
  (if (empty? items) ""
      (clojure.string/join
       "\n"
       (conj (mapv #(str "For want of a " %1 " the " %2 " was lost.")
                   items (rest items))
             (str "And all for the want of a " (first items) ".")))))
